SCDM for noncollinear systems

jaemolihm opened the pull request:
This PR implements the SCDM method for noncollinear systems. The change is almost localized in the SCDM-related part of the pw2wannier90 code, with additional removal of one io_error call in parameters.F90 that forbids SCDM with noncollinear.

Both version 6.3 and 6.4 of pw2wannier90 are updated.

The changes I have introduced in the SCDM-related part of the pw2wannier90.f90 code are threefold.

1) Fix small bugs in the current code: Davcio calls are moved outside the loop over bands, and an integer to double precision conversion problem has been fixed. The latter issue introduces a small but observable (~10<sup>-7</sup>) change in the output Amn values.

2) Use of slow Fourier transforms instead of FFT for calculating psi_nk at pivot positions. This is more efficient because the number of pivot points where the value of wavefunction should be calculated is much smaller than the number of FFT grid points. I have checked that the output Amn values are not affected by this modification.

3) Extension of the SCDM method to spinor wavefunctions. I have introduced a new subroutine compute_amn_with_scdm_spinor. There, pivot points for the pair of position and spin indices are calculated using QR factorization with column pivoting (QRCP).

Regards,
Jae-Mo Lihm
Seoul National University

jaemolihm committed:
Remove warning for spinor SCDM

jaemolihm committed:
Fix bugs in the SCDM part of pw2wannier90 v6.3: 1) davcio 2) int to double precision

jaemolihm committed:
Fix bugs in the SCDM part of pw2wannier90 v6.4: 1) davcio 2) int to double precision

jaemolihm committed:
Change pwscf/v6.3 compute_amn_with_scdm from using invfft to using slow Fourier transformation to compute psi_nk at pivot points.

jaemolihm committed:
Remove in pwscf/v6.3 one '#if defined(__MPI)' in compute_amn_with_scdm where invfft and gather_grid are no longer called, and mp_sum can be called without #if defined(__MPI).

jaemolihm committed:
Change pwscf/v6.4 compute_amn_with_scdm from using invfft to using slow Fourier transformation to compute psi_nk at pivot points. Also, removed the '#if defined(__MPI)' there.

jaemolihm committed:
Merge branch 'dev_scdm_slowFT' into dev_spinor_scdm

jaemolihm committed:
Implement SCDM for noncollinear wavefunction in pwscf/v6.3/pw2wannier90.f90

jaemolihm committed:
Change mp_pools to mp_global for v6.3/pw2wannier90.f90

jaemolihm committed:
Merge branch 'dev_scdm_slowFT' into dev_spinor_scdm

jaemolihm committed:
Change mp_pools to mp_global for v6.3/pw2wannier90.f90

jaemolihm committed:
Implement SCDM for noncollinear wavefunction in pwscf/v6.4/pw2wannier90.f90

jryates commented on pwscf/v6.3/pw2wannier90.f90:
```diff
@@ -3475,9 +3480,9 @@ SUBROUTINE compute_amn_with_scdm
-            rpos(lpt,1) = REAL(ipt)/dffts%nr1 
-            rpos(lpt,2) = REAL(jpt)/dffts%nr2 
-            rpos(lpt,3) = REAL(kpt)/dffts%nr3 
+            rpos(lpt,1) = DBLE(ipt)/DBLE(dffts%nr1)
```
I think it would be better to use REAL(ipt,DP) here, rather than use DBLE. To my mind DBLE is a more fortran77 way of handling precision.

jaemolihm replied:
Thank you for the comments. I have changed to them to REAL(*, DP). (Both for v6.3 and v6.4)

jryates commented on pwscf/v6.3/pw2wannier90.f90:
```diff
@@ -231,8 +232,9 @@ PROGRAM pw2wannier90
        'Non-collinear and gamma_only not implemented',1)
-  IF (noncolin.and.scdm_proj) CALL errore('pw2wannier90',&
-       'Non-collinear and SCDM not implemented',1)
+! jml
```
Be bold - just delete these lines! (we have the history in git)

jaemolihm replied:
I have done so :) (for both v6.3 and v6.4)

jaemolihm committed:
Changes regarding the comments on PR #277

VVitale commented:
@jaemolihm Thank you very much for your contribution. It would be nice if you could produce a tutorial example and a test case for the spin-orbit implementation of the SCDM at some point. Ideally, we would also ask for some documentation, both in the user guide and in the tutorial (see the guide line for contributors on https://github.com/wannier-developers/wannier90/wiki/ContributorsGuide).

jaemolihm commented:
@VVitale I am currently working on the tutorial, documentation, and the user guide. When I am complete, I will update them on this PR.
Regarding the test case, do you mean an automatic test in the test suite or an input/output example (as done in the solution_booklet)? As far as I understand, it is not trivial to make a test case for a pw2wannier feature since it requires a Quantum ESPRESSO run.

jaemolihm committed:
Update documentation on the SCDM for spinor wavefunctions: changelog, user guide, tutorial updated, and example31 added.

jaemolihm committed:
Update documentation on the SCDM for spinor wavefunctions: changelog, user guide, tutorial updated, and example31 added.

jaemolihm commented:
> @jaemolihm Thank you very much for your contribution. It would be nice if you could produce a tutorial example and a test case for the spin-orbit implementation of the SCDM at some point. Ideally, we would also ask for some documentation, both in the user guide and in the tutorial (see the guide line for contributors on https://github.com/wannier-developers/wannier90/wiki/ContributorsGuide).

I added example31 on SCDM for spinor wavefunctions and accordingly updated the tutorial. Also, I updated the user guide and the changelog.

hjunlee commented:
I think that for test purposes other useful and meaningful examples include the magnetic case, for instance bcc Fe, or the case with band inversion.

jaemolihm commented:
I have tested the code on magnetic systems (bcc iron and Mn3Ir, a noncollinear AFM) and a system with band inversion (BiTlSe2). The Wannierisation and band interpolation results are summarized in the attached pdf file.
[Wannier_SCDM_spinor.pdf](https://github.com/wannier-developers/wannier90/files/3357753/Wannier_SCDM_spinor.pdf)

giovannipizzi committed:
Some minor fixes to PR #277

These change slightly the text of the docs,
fix a compilation problem, and a couple of format
problems in the pw2wannier90 input file for example 31

giovannipizzi commented:
Dear @jaemolihm, thanks a lot for your very valuable contributions!

I think the PR is already in a very good shape.

I think it can be approved after the following two tasks:

1. approve my PR that I did against your repository/branch (these are a few typos in the text, an error preventing compilation in TeX, and a fix in the input file in the new example) - I found it easier to just do the changes than asking you to do them

2. creare a new folder `post-6.4` inside `pwscf`, and put in there a new version of the `pw2wannier90.f90` file, however carefully merging changes and bugfixes from the upstream Quantum ESPRESSO repository. In particular, [here](https://gitlab.com/QEF/q-e/commits/develop/PP/src/pw2wannier90.f90) you can see the recent history of the files, and in particular there are three commits that introduced a bugfix and some renaming:

  - [215a70b0](https://gitlab.com/QEF/q-e/commit/215a70b0c8edbc29c4b5f47d2386dfbe0eaf4411)
  - [f1ac07db](https://gitlab.com/QEF/q-e/commit/f1ac07dbe5dcb5c8f1901c74dec6c28e08987191)
  - [022652bf](https://gitlab.com/QEF/q-e/commit/022652bf29042d26ea0406af54afbb73d407f94f)

  It would be great if you could merge those changes with yours in a new file and put it in the folder I mention above, being careful if those changes should be replicated also in the parts of the code you are changing. 

As soon as these two tasks are done, I will merge this PR; afterwards, once this PR is merged, we will also do a PR against the main QE repo, before the file diverges again, and so that your contributions will be present in the next QE release.

giovannipizzi commented:
For reference, here is the diff of what has been done in the three commits mentioned above:
```diff
--- pw2wannier90-6.4.f90	2019-07-10 15:19:36.000000000 +0200
+++ pw2wannier90-qe-develop.f90	2019-07-10 15:32:36.000000000 +0200
@@ -1162,7 +1162,7 @@
    !
    ! Uses nkqs and index_sym from module pw2wan, computes rir
    !
-   USE symm_base,       ONLY : s, ftau, allfrac
+   USE symm_base,       ONLY : s, ft, allfrac
    USE fft_base,        ONLY : dffts
    USE cell_base,       ONLY : at, bg
    USE wannier,         ONLY : rir, read_sym
@@ -1174,8 +1174,8 @@
    INTEGER  , intent(in) :: nsym
    REAL(DP) , intent(in) :: sr(3,3,nsym), tvec(3,nsym)
    REAL(DP) :: st(3,3), v(3)
-   INTEGER, allocatable :: s_in(:,:,:), ftau_in(:,:)
-   !REAL(DP), allocatable:: ftau_in(:,:)
+   INTEGER, allocatable :: s_in(:,:,:)
+   REAL(DP), allocatable:: ft_in(:,:)
    INTEGER :: nxxs, nr1,nr2,nr3, nr1x,nr2x,nr3x
    INTEGER :: ikq, isym, i,j,k, ri,rj,rk, ir
    LOGICAL :: ispresent(nsym)
@@ -1189,7 +1189,7 @@
    nxxs = nr1x*nr2x*nr3x
    !
    !  sr -> s
-   ALLOCATE(s_in(3,3,nsym), ftau_in(3,nsym))
+   ALLOCATE(s_in(3,3,nsym), ft_in(3,nsym))
    IF(read_sym ) THEN
       IF(allfrac) THEN
          call errore("pw2wan_set_symm", "use_all_frac = .true. + read_sym = .true. not supported", 1)
@@ -1199,17 +1199,17 @@
          st = transpose( matmul(transpose(bg), transpose(sr(:,:,isym))) )
          s_in(:,:,isym) = nint( matmul(transpose(at), st) )
          v = matmul(transpose(bg), tvec(:,isym))
-         ftau_in(1,isym) = nint(v(1)*nr1)
-         ftau_in(2,isym) = nint(v(2)*nr2)
-         ftau_in(3,isym) = nint(v(3)*nr3)
+         ft_in(1,isym) = v(1)
+         ft_in(2,isym) = v(2)
+         ft_in(3,isym) = v(3)
       END DO
-      IF( any(s(:,:,1:nsym) /= s_in(:,:,1:nsym)) .or. any(ftau_in(:,1:nsym) /= ftau(:,1:nsym)) ) THEN
+      IF( any(s(:,:,1:nsym) /= s_in(:,:,1:nsym)) .or. any(ft_in(:,1:nsym) /= ft(:,1:nsym)) ) THEN
          write(stdout,*) " Input symmetry is different from crystal symmetry"
          write(stdout,*)
       END IF
    ELSE
       s_in = s(:,:,1:nsym)
-      ftau_in = ftau(:,1:nsym)
+      ft_in = ft(:,1:nsym)
    END IF
    !
    IF(.not. allocated(rir)) ALLOCATE(rir(nxxs,nsym))
@@ -1240,7 +1240,7 @@
             ENDDO
          ENDDO
    ENDDO
-   DEALLOCATE(s_in, ftau_in)
+   DEALLOCATE(s_in, ft_in)
 END SUBROUTINE pw2wan_set_symm
 
 !-----------------------------------------------------------------------
@@ -1669,7 +1669,6 @@
    !
    !
    IF(any_uspp) THEN
-      CALL init_us_1
       CALL allocate_bec_type ( nkb, nbnd, becp )
       IF (gamma_only) THEN
          call errore("compute_dmn", "gamma-only mode not implemented", 1)
@@ -2014,7 +2013,6 @@
    !
    !
    IF(any_uspp) THEN
-      CALL init_us_1
       CALL allocate_bec_type ( nkb, nbnd, becp )
       IF (gamma_only) THEN
          ALLOCATE ( rbecp2(nkb,nbnd))
@@ -2388,7 +2386,6 @@
    any_uspp = any(upf(1:ntyp)%tvanp)
 
    if (any_uspp) then
-      CALL init_us_1
       CALL allocate_bec_type ( nkb, nbnd, becp )
       ALLOCATE(be_n(nhm,2))
       ALLOCATE(be_m(nhm,2))
@@ -3059,7 +3056,6 @@
    !
    IF (any_uspp) THEN
       CALL allocate_bec_type ( nkb, n_wannier, becp)
-      CALL init_us_1
    ENDIF
    !
 
@@ -4142,7 +4138,7 @@
        ! 0-th Order
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! 1
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! 1
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
@@ -4150,21 +4146,21 @@
        ! 1st Order
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! x
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! x
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! y
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! y
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! z
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! z
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
@@ -4172,63 +4168,63 @@
        ! 2nd Order
        IF ( (abs(g_abc(1,igv) - 2.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! x^2
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! x^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! xy
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! xy
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) + 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! xy
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! xy
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! xz
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! xz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 1.d0 <= eps6) ) THEN ! xz
+            (abs(g_abc(3,igv) + 1.d0) <= eps6) ) THEN ! xz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 2.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! y^2
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! y^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! yz
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! yz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 1.d0 <= eps6) ) THEN ! yz
+            (abs(g_abc(3,igv) + 1.d0) <= eps6) ) THEN ! yz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 2.d0 <= eps6) ) THEN ! z^2
+            (abs(g_abc(3,igv) - 2.d0) <= eps6) ) THEN ! z^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
@@ -4236,133 +4232,133 @@
        ! 3rd Order
        IF ( (abs(g_abc(1,igv) - 3.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! x^3
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! x^3
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 2.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! x^2y
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! x^2y
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 2.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) + 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! x^2y
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! x^2y
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 2.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! x^2z
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! x^2z
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 2.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 1.d0 <= eps6) ) THEN ! x^2z
+            (abs(g_abc(3,igv) + 1.d0) <= eps6) ) THEN ! x^2z
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 2.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! xy^2
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! xy^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) + 2.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! xy^2
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! xy^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! xyz
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! xyz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 1.d0 <= eps6) ) THEN ! xyz
+            (abs(g_abc(3,igv) + 1.d0) <= eps6) ) THEN ! xyz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) + 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! xyz
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! xyz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) + 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 1.d0 <= eps6) ) THEN ! xyz
+            (abs(g_abc(3,igv) + 1.d0) <= eps6) ) THEN ! xyz
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 2.d0 <= eps6) ) THEN ! xz^2
+            (abs(g_abc(3,igv) - 2.d0) <= eps6) ) THEN ! xz^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 1.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 2.d0 <= eps6) ) THEN ! xz^2
+            (abs(g_abc(3,igv) + 2.d0) <= eps6) ) THEN ! xz^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 3.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 0.d0 <= eps6) ) THEN ! y^3
+            (abs(g_abc(3,igv) - 0.d0) <= eps6) ) THEN ! y^3
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 2.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 1.d0 <= eps6) ) THEN ! y^2z
+            (abs(g_abc(3,igv) - 1.d0) <= eps6) ) THEN ! y^2z
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 2.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 1.d0 <= eps6) ) THEN ! y^2z
+            (abs(g_abc(3,igv) + 1.d0) <= eps6) ) THEN ! y^2z
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 2.d0 <= eps6) ) THEN ! yz^2
+            (abs(g_abc(3,igv) - 2.d0) <= eps6) ) THEN ! yz^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and.&
             (abs(g_abc(2,igv) - 1.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) + 2.d0 <= eps6) ) THEN ! yz^2
+            (abs(g_abc(3,igv) + 2.d0) <= eps6) ) THEN ! yz^2
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
        ENDIF
        IF ( (abs(g_abc(1,igv) - 0.d0) <= eps6) .and. &
             (abs(g_abc(2,igv) - 0.d0) <= eps6) .and. &
-            (abs(g_abc(3,igv)) - 3.d0 <= eps6) ) THEN ! z^3
+            (abs(g_abc(3,igv) - 3.d0) <= eps6) ) THEN ! z^3
            num_G(mpime+1) = num_G(mpime+1) + 1
            ig_idx(num_G(mpime+1))=igv
            CYCLE
```

jaemolihm committed:
Merge pull request #1 from giovannipizzi/fixes_to_pr_277

Some minor fixes to PR #277

jaemolihm committed:
Merge remote branch 'origin/dev_spinor_scdm' into dev_spinor_scdm

Conflicts:
	doc/tutorial/tutorial.tex
	doc/user_guide/projections.tex
	doc/user_guide/wannier-pp.tex
	examples/example31/Pt.pw2wan

jaemolihm committed:
Create pwscf/post-6.4, put the new version of pw2wannier90 from develop branch of Quantum ESPRESSO, and merge the spinor-SCDM-related changes to it.

jaemolihm commented:
Dear @giovannipizzi, thanks for the feedback!

I have done the two tasks.
Regarding the second task, I have merged my changes to the new version of pw2wannier90.f90 in post-6.4. None of them were related with the part of the code that I am changing (the SCDM part). I have checked that it compiles and runs correctly.

giovannipizzi reviewed (approved):
Thanks a lot! I tested the code in v6.4 and it works. Also thanks for addressing the support of future new versions of QE. I will now merge this and make a PR against the QE repository.

giovannipizzi commented:
For future reference, the merge request to the Quantum ESPRESSO GitLab repo is [here](https://gitlab.com/QEF/q-e/merge_requests/522).

jaemolihm commented:
For future reference, I attached the input files, pseudopotentials, and Wannier90 output for noncollinear SCDM, which were used to generate the note attached above ("Wannier_SCDM_spinor.pdf").

[scdm_noncollinear_input.tar.gz](https://github.com/wannier-developers/wannier90/files/10430219/scdm_noncollinear_input.tar.gz)
